Add SMUFF support

GMagician opened the pull request:
Changed MMU definition to a per model basis
Added SMUFF support (that actually let user to choose from 2 up to 12 colors)

**~~SMUFF needs more coding to be really working, feedback needed to x model basis approach~~**

To know more about SMUFF: https://sites.google.com/view/the-smuff/home

GMagician commented:
I found that code does't compile if `MMU2_MENUS` is not enabled on MMU2S.
I don't know how mmu2 should work but I see that it seems that Tx/T? can't be handled it no lcd menu...someone can check how this should work?

GMagician committed:
Add SMUFF support

- Changed MMU definition to a per model basis
- Added SMUFF support (that will actually only expand multicolor up to 15 colors)

thinkyhead committed:
Update platformio.ini

thinkyhead committed:
Config tweaks

thinkyhead committed:
Fix up config

thinkyhead committed:
Set HAS_PRUSA_MMU2 for S also

thinkyhead committed:
Adjust formatting

thinkyhead commented:
If an extruder is emulated, Marlin can support any number of tools. So the `MAX_EXTRUDERS` value should just be updated to 15 if a SMUFF is connected.

thinkyhead committed:
Set MAX_EXTRUDERS to 15 for SMUFF

thinkyhead committed:
Drop some MAX_EXTRUDERS checks

thinkyhead committed:
MMU2S is also MMU2

thinkyhead committed:
Misc adjustments

thinkyhead commented:
A controller is needed for the MMU2 tool-changing because there is a call to `mmu2_choose_filament` which asks for user feedback. If not for this, you would not need one of the supported controllers. It seems to only affect the "special" version of `tool_change`.

thinkyhead committed:
Towards MMU2 without LCD

GMagician commented:
Ok, than if it's required to have lcd some functions then some fix have to be done to let mmu2 compile without lcd, I'll address them (if not already done)

GMagician committed:
Shrink some remarks

GMagician committed:
Shrink remarks message

GMagician commented on Marlin/src/inc/Conditionals_LCD.h:
```diff
@@ -539,17 +567,17 @@
-// No inactive extruders with MK2_MULTIPLEXER or SWITCHING_NOZZLE
-#if EITHER(MK2_MULTIPLEXER, SWITCHING_NOZZLE)
+// No inactive extruders with SWITCHING_NOZZLE or Průša MMU1
+#if ENABLED(SWITCHING_NOZZLE) || HAS_PRUSA_MMU1
```
@thinkyhead I'm wondering what is better between `#if ENABLED(SWITCHING_NOZZLE) || HAS_PRUSA_MMU1` and `#if EITHER(SWITCHING_NOZZLE, HAS_PRUSA_MMU1)`. Both seems to compile correctly but HAS_PRUSA_MMU1 is not a directly activated option (and there is a mix in code that I don't like, I would prefer a uniform approach)

thinkyhead replied:
The function of these `ENABLED` (etc.) macros is to digest any config switches and any switch-like conditionals, converting them into ones, zeros, and boolean operators. These switch-like conditionals can be used naked in pre-processor directives, as you would expect, but if you want to use them in code, for example a simple `if (HAS_PRUSA_MMU1)` then it won't work. You must wrap it one of the `ENABLED` macros to use it in code. It's a matter of stylistic choice whether to also apply them in the pre-compiler directives where they are not required.

GMagician commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -874,6 +874,31 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
+ */
+#if HAS_PRUSA_MMU2
+  #if EXTRUDERS != 5
+    #undef SINGLENOZZLE
```
I think this is senseless (the undef) since code will not compile, I think I missed this before

thinkyhead replied:
I put this #undef here to suppress a sanity-check error message that makes reference to "`SINGLENOZZLE`" because the `SINGLENOZZLE` option is only enabled as a condition for the MMU, and the user does not need to enable it themselves.

thinkyhead replied:
There's a good reason for it, but GitHub ate my last reply and I don't feel like typing it out again.

GMagician commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -874,6 +874,31 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
+    #error "PRUSA_MMU2S requires FILAMENT_RUNOUT_SENSOR. Enable it to continue."
+  #elif ENABLED(MMU_EXTRUDER_SENSOR) && DISABLED(FILAMENT_RUNOUT_SENSOR)
+    #error "MMU_EXTRUDER_SENSOR requires FILAMENT_RUNOUT_SENSOR. Enable it to continue."
+  #elif HAS_PRUSA_MMU2S && !HAS_LCD_MENU
```
Is this really need anymore?

GMagician commented:
@thinkyhead I'm not sure if it worth to add a subfolder to mmu for mmu1 and mmu2. smuff has it's own protocol and once I have one I'll see if such protocol gives some benefits over mmu2 one, if so more files will appear here mixing up things. On the other way increasing path depth may give some "path too long" issues. Any suggestion?

GMagician committed:
Reduce smuff supported extruders

Due to some changes that will be done soon on smuff firmware, next release, max supported extruders will be reduced

GMagician committed:
Reduce smuff supported colors

Next smuff firmware release new limit

GMagician committed:
Remove senseless #undef

thinkyhead committed:
Restore senseless #undef

GMagician commented:
@thinkyhead why have you readded `#undef`? compile will fail due to `#error` and MMU needs SINGLENOZZLE

thinkyhead committed:
human-readability

thinkyhead committed:
Future units

thinkyhead committed:
Clean up, permit "Tc" with no LCD

GMagician committed:
Use common way to test MMU options in precompiler conditions

GMagician committed:
Use TERN_ instead of #if

GMagician commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -874,6 +874,31 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
+    #error "PRUSA_MMU2S requires FILAMENT_RUNOUT_SENSOR. Enable it to continue."
+  #elif ENABLED(MMU_EXTRUDER_SENSOR) && DISABLED(FILAMENT_RUNOUT_SENSOR)
+    #error "MMU_EXTRUDER_SENSOR requires FILAMENT_RUNOUT_SENSOR. Enable it to continue."
+  #elif HAS_PRUSA_MMU2S && !HAS_LCD_MENU
```
Is this sanity check still needed after last changes (line 890)?

GMagician replied:
Ok removed

GMagician committed:
Use SMUFF developper nickname

Grogyan commented:
Bigtree Tech have announced today that they are releasing a new control board for the Prusa MMU2, from the initial info, this board is well suited for the SMuFF as well

GMagician commented:
@Grogyan good to know...I'll check when more info are available

GMagician committed:
Some fixes

- Add #20140 unsigned warning fix
- Fix non inluded cnf when no mmu2 menus

GMagician committed:
Remove no more valid sanity check

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19912

thinkyhead committed:
Remove extra marker

thinkyhead committed:
Move SNMM checks up

thinkyhead committed:
Fix some tests

thinkyhead committed:
and this test too

thinkyhead committed:
Test seems intended for 5 sensors

thinkyhead committed:
Clarify sanity error with MMU2

thinkyhead committed:
Fixes, more to come

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19912

thisiskeithb commented:
@GMagician: Did you see that the project has been discontinued? https://sites.google.com/view/the-smuff/home

GMagician commented:
Yes I saw. V5 has been discontinued but V4.75 is still present on thingiverse
